src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.cpp

```diff
@@ -377,33 +273,161 @@ namespace argos {
+      CVector3 cLeft = m_sActivePlacement.Up;
+      cLeft.CrossProduct(cForward).Normalize();
+      /* Calculate motion sensitivity */
+      Real fMotionSensitivity = MOVE_GAIN * ::exp(m_sActivePlacement.LensFocalLength);
+      /* Apply translation */
+      m_sActivePlacement.Position += 
+         cForward * (fMotionSensitivity * n_forwards_backwards) +
+         cLeft * (fMotionSensitivity * n_sideways) +
+         cUp * (fMotionSensitivity * n_up_down);
+      m_sActivePlacement.Target = m_sActivePlacement.Position;
+      m_sActivePlacement.Target += cForward;
+   }
+
+   /****************************************/
+   /****************************************/
+
+   void CQTOpenGLCamera::Interpolate(UInt32 un_start_placement,
+                                     UInt32 un_end_placement,
+                                     Real f_time_fraction) {
+      const SPlacement& sStart = m_arrPlacements[un_start_placement];
+      const SPlacement& sEnd = m_arrPlacements[un_end_placement];
+      /* Linear interpolate position vector */
+      m_sActivePlacement.Position = (sEnd.Position - sStart.Position);
+      m_sActivePlacement.Position *= f_time_fraction;
+      m_sActivePlacement.Position += sStart.Position;
+      /* Linear interpolate look at vector */
+      m_sActivePlacement.Target = (sEnd.Target - sStart.Target);
+      m_sActivePlacement.Target *= f_time_fraction;
+      m_sActivePlacement.Target += sStart.Target;
+      /* Linear interpolate up vector */
+      m_sActivePlacement.Up = (sEnd.Up - sStart.Up);
+      m_sActivePlacement.Up *= f_time_fraction;
+      m_sActivePlacement.Up += sStart.Up;
+      /* Linear interpolate the focal length */
+      m_sActivePlacement.LensFocalLength = (sEnd.LensFocalLength - sStart.LensFocalLength);
+      m_sActivePlacement.LensFocalLength *= f_time_fraction;
+      m_sActivePlacement.LensFocalLength += sStart.LensFocalLength;
+      /* Calculate the Y field-of-view */
+      m_sActivePlacement.CalculateYFieldOfView();
    }
 
    /****************************************/
    /****************************************/
 
    void CQTOpenGLCamera::UpdateTimeline() {
-      /* If no timeline, nothing to do */
-      if(m_listTimeline.empty()) return;
-      /* Get current timestep */
-      UInt32 unStep = CSimulator::GetInstance().GetSpace().GetSimulationClock();
-      /* Search for active timeline item */
-      std::list<STimelineItem*>::iterator it = m_listTimeline.begin();
-      /* Is the first item potentially active? */
-      if((*it)->Step <= unStep) {
-         /* Yes, save it and search for subsequent items */
-         STimelineItem* psTimelineItem = *it;
-         ++it;
-         while(it != m_listTimeline.end() && (*it)->Step <= unStep) {
-            psTimelineItem = *it;
-            ++it;
+      if(!m_bEnableTimeline)
+         return;
+      /* initialize two iterators to the first keyframe */
+      std::vector<SKeyframe>::const_iterator itCurrent = 
```

Comment:
This function seems too long. Consider breaking into smaller functions to improve readability:

- Handle looping case
- Handle non-looping case

Reply:
This function is too long and I think there are better ways of handling this logic. Although for this PR, I am focusing mostly on correctness and functionality. I will leave further refactoring and improving this logic to a future PR...